### PR TITLE
Image generation through Dalle

### PR DESCRIPTION
**Pull request by clemsau**
This should allow image generation.

Maybe the user could also choose the size of the image to generate. I am open to modifications. Let me know :smiley:

### CODE REVIEW

**Comment by Iambabyninja** [comment]
Why error?
An exception was raised while handling an update
update = {
  "update_id": 563864453,
  "message": {
    "chat": {
      "id": [5973962547](tel:5973962547),
      "type": "private",
      "last_name": "Funny",
      "first_name": "Not",
      "username": "life_Amem"
    },
    "text": "/imagine cat",
    "group_chat_created": false,
    "entities": [
      {
        "length": 8,
        "type": "bot_command",
        "offset": 0
      }
    ],
    "message_id": 5114,
    "delete_chat_photo": false,
    "date": [1679244930](tel:1679244930),
    "supergroup_chat_created": false,
    "channel_chat_created": false,
    "from": {
      "is_bot": false,
      "username": "life_Amem",
      "first_name": "Not",
      "last_name": "Funny",
      "id": [5973962547](tel:5973962547),
      "language_code": "ru"
    }
  }
}

Traceback (most recent call last):
  File "/usr/local/lib/python3.8/site-packages/telegram/ext/_application.py", line 1104, in process_update
    await coroutine
  File "/usr/local/lib/python3.8/site-packages/telegram/ext/_handler.py", line 141, in handle_update
    return await self.callback(update, context)
  File "bot/bot.py", line 389, in image_generation_handle
    image_url = await openai_utils.generate_image(caption)
  File "/code/bot/openai_utils.py", line 177, in generate_image
    r = await openai.Image.create(prompt=prompt, n=1, size="256x256")
TypeError: object OpenAIObject can't be used in 'await' expression

**Comment by clemsau** [comment]
Haven't ran any tests on it, I'll test manually and fix the bugs @Iambabyninja

**Comment by Rxmsey** [comment]
whats ur discord?

**Comment by clemsau** [comment]
Just got it to work

image

Will upload the patch soon

**Comment by clemsau** [comment]
Just updated, working well. If ppl want to try it: `git clone -b image-generation git@github.com:clemsau-forks/chatgpt_telegram_bot.git`

**Comment by clemsau** [comment]
I just pushed other changes because @Rxmsey showed me that some requests could violate openai's policy (e.g `Trump crying`). Whenever it happens, the bot will reply `Something went wrong during image generation. It's possible that your request not allowed by openai's safety system`

**Comment by Iambabyninja** [comment]
Thank you!
It's work good 4 me ❤️

File: bot/bot.py [hunk 1]
```diff
@@ -33,6 +33,11 @@
 ⚪ /retry – Regenerate last bot answer
 ⚪ /new – Start new dialog
 ⚪ /mode – Select chat mode
+<<<<<<< HEAD
```
Comment (Tuarisa) [inline bot/bot.py:36]: Hi!

It seems that there are some unresolved merge conflict markers in the HELP_MESSAGE section of the code.

**Comment by Tuarisa** [comment]
Hello. Thank you for your work. I created my own fork and applied your patch in it, and it works perfectly.

I would also like for this patch to be available to everyone, but the pull request has conflicts with the current branch. Additionally, I think it would be a good idea to add a configuration setting that would allow the use of DALL-E. Since this feature consumes a lot of resources, not all public bot owners would want to include it by default due to faster credit consumption.

Let me know if you need help with this, or I can create my own PR based on yours with the suggested changes. Thank you once again for your work.

**Comment by clemsau** [comment]
@Tuarisa I am not looking to put further work in this PR tbh. The conflicts are caused by the many changes that happened since I pushed this PR.
Happy to hear it works well for you :smiley: ! 
And yeah, If you think this needs further changes or additional functionalities (the option to disable dalle seems to be needed indeed), you can push another PR with additional changes on the basis of this PR. Or even try to do it better :smiley:

**Comment by jorgik1** [comment]
Hello, I've had an error in the line: `db.set_user_attribute(user_id, "n_used_tokens", config.dalle_price_per_image + db.get_user_attribute(user_id, "n_used_tokens"))`

```
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/site-packages/telegram/ext/_application.py", line 1104, in process_update
    await coroutine
  File "/usr/local/lib/python3.8/site-packages/telegram/ext/_handler.py", line 141, in handle_update
    return await self.callback(update, context)
  File "bot/bot.py", line 339, in image_generation_handle
    config.dalle_price_per_image + db.get_user_attribute(user_id, "n_used_tokens"))
TypeError: unsupported operand type(s) for +: 'float' and 'dict'
```

**Comment by clemsau** [comment]
@jorgik1 Somehow `db.get_user_attribute(user_id, "n_used_tokens")` returned you a dict :thinking: I have no idea why this happened

**Comment by jorgik1** [comment]
I think it's related to:
`    # back compatibility for n_used_tokens field
    n_used_tokens = db.get_user_attribute(user.id, "n_used_tokens")
    if isinstance(n_used_tokens, int):  # old format
        new_n_used_tokens = {
            "gpt-3.5-turbo": {
                "n_input_tokens": 0,
                "n_output_tokens": n_used_tokens
            }
        }
        db.set_user_attribute(user.id, "n_used_tokens", new_n_used_tokens)`

**Review by hkvn — APPROVED** [review]

**Comment by clemsau** [comment]
This was used in #245 
I am closing the PR

**Pull request closed without merging**